server/src/test/java/io/crate/planner/optimizer/costs/PlanStatsTest.java

```diff
@@ -0,0 +1,144 @@
+
+public class PlanStatsTest extends CrateDummyClusterServiceUnitTest {
+
+    public void test_collect() throws Exception {
+        SQLExecutor e = SQLExecutor.builder(clusterService)
+            .addTable("create table a (x int)")
+            .build();
+
+        DocTableInfo a = e.resolveTableInfo("a");
+
+        var x = e.asSymbol("x");
+        var source = new Collect(new DocTableRelation(a),
+                                 List.of(x),
+                                 WhereClause.MATCH_ALL,
+                                 1L,
+                                 DataTypes.INTEGER.fixedSize());
+        var memo = new Memo(source);
+        TableStats tableStats = new TableStats();
+        tableStats.updateTableStats(Map.of(a.ident(), new Stats(1, 1, Map.of())));
+        PlanStats planStats = new PlanStats(tableStats, memo);
+        var result = planStats.apply(source);
+        assertThat(result.numDocs()).isEqualTo(1L);
+    }
+
+    public void test_group_reference() throws Exception {
+        SQLExecutor e = SQLExecutor.builder(clusterService)
+            .addTable("create table a (x int)")
+            .build();
+
+        DocTableInfo a = e.resolveTableInfo("a");
+
+        var x = e.asSymbol("x");
+        var source = new Collect(new DocTableRelation(a),
+                                 List.of(x),
+                                 WhereClause.MATCH_ALL,
+                                 1L,
+                                 DataTypes.INTEGER.fixedSize());
+        var groupReference = new GroupReference(1, source.outputs(), Set.of());
+        var memo = new Memo(source);
+        TableStats tableStats = new TableStats();
+        tableStats.updateTableStats(Map.of(a.ident(), new Stats(1, 1, Map.of())));
+        PlanStats planStats = new PlanStats(tableStats, memo);
+        var result = planStats.apply(groupReference);
+        assertThat(result.numDocs()).isEqualTo(1L);
+    }
+
+    public void test_tree_of_operators() throws Exception {
+        SQLExecutor e = SQLExecutor.builder(clusterService)
+            .addTable("create table a (x int)")
+            .build();
+
+        DocTableInfo a = e.resolveTableInfo("a");
+
+        var x = e.asSymbol("x");
+        var source = new Collect(new DocTableRelation(a),
+                                 List.of(x),
+                                 WhereClause.MATCH_ALL,
+                                 10L,
+                                 DataTypes.INTEGER.fixedSize());
+        TableStats tableStats = new TableStats();
+        tableStats.updateTableStats(Map.of(a.ident(), new Stats(10L, 1, Map.of())));
+        var limit = new Limit(source, Literal.of(5), Literal.of(0));
+
+        var memo = new Memo(limit);
+        PlanStats planStats = new PlanStats(tableStats, memo);
+        var result = planStats.apply(limit);
+        assertThat(result.numDocs()).isEqualTo(5L);
+    }
+
+    public void test_update_table_stats() throws Exception {
```

Comment:
What's this test having to do with updating table stats?  PlanStats always just uses the provided TableStats.

Reply:
The numbers in the tablestats change.

Reply:
👍